### PR TITLE
DIV-4571 Filter cases by user divorce role

### PR DESCRIPTION
**Pull request by qzhou-hmcts**
# Description
https://tools.hmcts.net/jira/browse/DIV-4571
Filtered users based on Petitioner / Respondent role.
Filtered Amended cases on  return DN endpoint

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

**Test Configuration**:

* Hardware:
* O/S and version:
* JDK:

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/integrationTest/java/uk/gov/hmcts/reform/divorce/petition/CcdRetrieveAosCaseTest.java [hunk 1]
```diff
@@ -118,32 +119,32 @@ public void givenAosStartedCaseInCcd_whenRetrieveAosCase_thenReturnTheCase() thr
     @Test
     public void givenMultipleAosStartedAndNoAosCompletedCaseInCcd_whenRetrieveAosCase_thenReturnMultipleChoice()
             throws Exception {
-        final String userToken = getUserToken();
+        final UserDetails userDetails = getUserDetails();
 
-        createACaseUpdateStateAndReturnTheCase(userToken, TEST_AOS_RESPONDED_EVENT).path("id");
-        createACaseUpdateStateAndReturnTheCase(userToken, TEST_AOS_RESPONDED_EVENT).path("id");
+        createACaseUpdateStateAndReturnTheCase(userDetails, TEST_AOS_RESPONDED_EVENT).prettyPrint();
```
Comment (hosainnet) [inline src/integrationTest/java/uk/gov/hmcts/reform/divorce/petition/CcdRetrieveAosCaseTest.java]: Not sure we should leave prettyPrint here

File: src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdRetrievalServiceImpl.java [hunk 2]
```diff
@@ -126,15 +138,36 @@ public CaseDetails retrieveCaseById(String authorisation, String caseId) {
...
+        return cases.stream()
+            .filter(caseDetails -> userHasSpecifiedRole(caseDetails, user.getEmail(), role))
+            .collect(Collectors.toList());
+
+    }
+
+    private boolean userHasSpecifiedRole(CaseDetails caseDetails, String userEmail, DivCaseRole role) {
+
+        if (role == null) {
+            return false;
```
Comment (hosainnet) [inline src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdRetrievalServiceImpl.java]: Should probably thought an exception here?

File: src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdRetrievalServiceImpl.java [hunk 3]
```diff
@@ -126,15 +138,36 @@ public CaseDetails retrieveCaseById(String authorisation, String caseId) {
...
+        }
+
+        switch (role) {
+            case PETITIONER:
+                return  userEmail.equalsIgnoreCase((String) caseDetails.getData().get(D8_PETITIONER_EMAIL));
+            case RESPONDENT:
+                return  userEmail.equalsIgnoreCase((String) caseDetails.getData().get(CO_RESP_EMAIL_ADDRESS))
+                    || userEmail.equalsIgnoreCase((String) caseDetails.getData().get(RESP_EMAIL_ADDRESS));
```
Comment (hosainnet) [inline src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdRetrievalServiceImpl.java]: Do we not want to have a more explicit CO_RESPONDENT role?

File: src/integrationTest/java/uk/gov/hmcts/reform/divorce/petition/CcdRetrieveAosCaseTest.java (hunk 1 above)
Comment (qzhou-hmcts) [reply src/integrationTest/java/uk/gov/hmcts/reform/divorce/petition/CcdRetrieveAosCaseTest.java]: I'll delete it, I added while I was debugging  and I forgot to remove it cheers.

Comment (hosainnet) [reply src/integrationTest/java/uk/gov/hmcts/reform/divorce/petition/CcdRetrieveAosCaseTest.java]: Probably good to add as a message for failures actually like https://github.com/hmcts/div-case-maintenance-service/pull/112#discussion_r271284106

File: src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdRetrievalServiceImpl.java (hunk 3 above)
Comment (qzhou-hmcts) [reply src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdRetrievalServiceImpl.java]: the problem here is you don't know what is the role you are using unless you check the email field as  Respondent and CoRespondent use the same endpoint to retrieve the case

**Pull request merged**